### PR TITLE
One line cr bot

### PR DESCRIPTION
**Pull request by nmanthey**
*Issue #, if available:*

This series fixes the invocation of Infer, to allow using Infer v1.0.0. Furthermore, Infer and CppCheck analysis can be controlled via environment variables.

The documentation of the tool is improved as well, including describing how it works, providing examples and explaining environment variables.

The one-line-cr-bot.sh script is added to the repository as well, which allows to analyze a git commit series and presents the set of introduced defects. See e.g. https://github.com/nmanthey/ktf/runs/1271790410?check_suite_focus=true for testing done as part of a github workflow.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by nmanthey** [comment]
This commit series includes the #12  series as well, i.e. the major difference here is the addition of CppCheck environment variable usage, as well as one-line-cr-bot.sh

File: configuration/doc/EXAMPLES.md [hunk 1]
```diff
@@ -60,3 +60,23 @@ x86_64-unknown-linux-gcc can be used, as well as other cross-compilers. To use
...
+As Infer calls are happening inside one-line-scan, it might be hard to add
+additional analysis options to Infer. The below environment variable
+INFER_ANALYSIS_EXTRA_ARGS can be used to forward analysis options.
+
+The following commands can be used to run Infer on a project with compiler
+'new-compiler', and with extra analysis options '--bufferoverrun'.
+
+  one-line-scan -o OLS --use-existing --no-gotocc --infer --no-analysis -- cmake
+  INFER_ANALYSIS_EXTRA_ARGS="--bufferoverrun" \
+      OLS_TARGET_COMPILER="my-compiler" \
```
Comment (tautschnig) [inline configuration/doc/EXAMPLES.md:81]: How about using `export` here to break up the very long line? At least I found this a bit hard to parse.

File: configuration/doc/README.md [hunk 2]
```diff
@@ -27,11 +27,121 @@ plain backend, which only tracks calls to the compiler, but does not act upon
...
-configuration/doc/EXAMPLES.md
+The available parameters of the tool can be seen when using the --help flag:
+
+  one-line-scan --help
+
+More invocations of one-line-scan are presented in the file doc/EXAMPLES.md
+
+By default, the wrapper for CBMC is enabled. Furthermore, analysis is enabled by
+default. To disable CBMC, use --no-gotocc, and to disable analysis, use the
+parameter --no-analysis.
```
Comment (tautschnig) [inline configuration/doc/README.md:40]: Just a thought: should we actually make "plain" the default?

File: configuration/doc/README.md [hunk 3]
```diff
@@ -27,11 +27,121 @@ plain backend, which only tracks calls to the compiler, but does not act upon
...
+
+## How It Works
+
+The tool one-line-scan places wrappers for common compilers like gcc, clang, as
+well as g++ and clang++ in a new directory. If the parameter --suffix $SUFFIX
+or --prefix $PREFIX is used, the wrapper is also created for binaries of the
+name ${PREFIX}compiler and compiler${SUFFIX}. This allows to also track compiler
```
Comment (tautschnig) [inline configuration/doc/README.md:47]: track compiler*s*

File: configuration/doc/README.md [hunk 4]
```diff
@@ -27,11 +27,121 @@ plain backend, which only tracks calls to the compiler, but does not act upon
...
+like gcc-9, or x86_64-unknown-linux-gnu-gcc.
+
+After all wrappers have been placed in the directory, the environment variable
+PATH is prepended with this directory. Hence, any attempt to spot the compiler
+via the PATH variable will result in picking up the wrapper, instead of the
+actual compiler. This allow us the take action on compiler invocations.
```
Comment (tautschnig) [inline configuration/doc/README.md:53]: allow*s*

File: configuration/doc/README.md [hunk 5]
```diff
@@ -27,11 +27,121 @@ plain backend, which only tracks calls to the compiler, but does not act upon
...
+  one-line-scan --plain --no-gotocc -o OLS --trunc-existing -- which gcc
+
+With this mechanism, compiler calls based on their absolute path cannot be
+intercepted. Consequently, analysis will not detect these calls. To still be
+able to get these calls, one-line-scan should be used during the configuration
+with "./configure" or "cmake", and next should re-use the created directory for
+the actual compilation. This way, the wrapper will be used during configuration.
+See call examples in configuration/doc/EXAMPLES.md or more details below.
+
+## How to Use For
```
Comment (tautschnig) [inline configuration/doc/README.md:67]: Maybe add "..." or instead say "Usage Scenarios"

File: configuration/doc/README.md [hunk 6]
```diff
@@ -27,11 +27,121 @@ plain backend, which only tracks calls to the compiler, but does not act upon
...
+
+### Different Build Systems
+
+This section explains the usage of one-line scan for the build systems using
```
Comment (tautschnig) [inline configuration/doc/README.md:71]: one-line*-*scan

File: configuration/doc/README.md [hunk 7]
```diff
@@ -27,11 +27,121 @@ plain backend, which only tracks calls to the compiler, but does not act upon
...
+plain compilation, make, configure and make, as well as cmake, and give a brief
+example on what should be done.
```
Comment (tautschnig) [inline configuration/doc/README.md:73]: Instead of "cmake, and give a brief example on what should be done." I'd say: "cmake. For each of them, a brief example is provided."

File: configuration/doc/README.md [hunk 8]
```diff
@@ -27,11 +27,121 @@ plain backend, which only tracks calls to the compiler, but does not act upon
...
+Make typically uses compilers directly, without an absolute path -- depending on
+the actually used Makefile. Hence, the following command should be able to get
+the used compiler calls:
+
+  one-line-scan -o OLS --trunc-existing --no-gotocc --plain -- make
+
+#### Configure and Make
+
+Make typically uses compilers directly, without an absolute path. However, a
+preceeding configure call might set the absolute path based on the tools that
```
Comment (tautschnig) [inline configuration/doc/README.md:95]: s/preceeding/preceding/

File: one-line-cr-bot.sh [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+# Copyright (C) 2020 Amazon.com, Inc. or its affiliates.
+# Author: Norbert Manthey <nmanthey@amazon.de>
+#
+# Script to report errors/warnings/static code analysis output for a git commit series
+#
+# Tools currently used: infer, cppcheck
+# To potentially add: gcc-10 fanalyzer, cbmc
+#
+# If cppcheck or infer are not present in PATH, the tools can be obtained, and
+# build from their respective source in the www, all in a temporary folder that
```
Comment (tautschnig) [inline one-line-cr-bot.sh:12]: s/build/built/

File: one-line-cr-bot.sh [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+# will be thrown away after the invocation. If you plan to use this script
+# multiple time, consider installing these tools natively! To use this feature,
```
Comment (tautschnig) [inline one-line-cr-bot.sh:14]: s/time/times/

File: one-line-cr-bot.sh [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+# use the flag -I, or assign the value "true" to the environment variable
+# INSTALL_MISSING. Note, installation from the www might fail due to
+# unavailability.
+#
+# To customize the analysis of the used tools, check the documentation for the
+# backends of the one-line-scan tool. Configuration via environment variables is
+# possible.
+
+# Variables that can be influenced via environment, or CLI
```
Comment (tautschnig) [inline one-line-cr-bot.sh:23]: s/influenced/set/

File: one-line-cr-bot.sh [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+declare -A RUN_TOOL
+RUN_TOOL["INFER"]="true"
+RUN_TOOL["CPPCHECK"]="true"
+
+#
+# DO NOT TOUCH THE BELOW
+#
+readonly PROG=$(basename "$0")
+readonly SCRIPT_DIR=$(readlink -f "$(dirname "$0")")
+readonly REPOSITORY_DIR=${PWD}
```
Comment (tautschnig) [inline one-line-cr-bot.sh:50]: Is there a particular reason to use `${PWD}` instead of `$PWD`?

File: one-line-cr-bot.sh [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+
+TMP_DATA_FOLDER=""
+
+usage() {
+    cat <<EOF
+One-Line-CR Bot
+
+This script allows to run code analysis based on One-Line-Scan against a given
+code base, and eventually print the diff in defects wrt a base commit. The
```
Comment (tautschnig) [inline one-line-cr-bot.sh:59]: "eventually" makes it sound like this could take forever. How about "code base. Using these results, the user can choose to print the diff in defects in comparison to those found for a base commit."

File: one-line-cr-bot.sh [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+commands to be used, as well as the commits to be used for comparison, can be
+specified via environment or CLI parameters, where CLI will override the
+environment.
+
+Tools currently used: ${SUPPORTED_TOOLS[*]}
+
+If cppcheck or infer are not present in PATH, the tools can be obtained, and
+build from their respective source in the www, all in a temporary folder that
```
Comment (tautschnig) [inline one-line-cr-bot.sh:67]: s/build/built/

File: one-line-cr-bot.sh [hunk 15]
```diff
@@ -0,0 +1,531 @@
...
+will be thrown away after the invocation. If you plan to use this script
+multiple time, consider installing these tools natively! To use this feature,
```
Comment (tautschnig) [inline one-line-cr-bot.sh:69]: s/time/times/

File: one-line-cr-bot.sh [hunk 16]
```diff
@@ -0,0 +1,531 @@
...
+use the flag -I, or assign the value "true" to the environment variable
+INSTALL_MISSING. Note, installation from the www might fail due to
+unavailability.
+
+To customize the analysis of the used tools, check the documentation for the
```
Comment (tautschnig) [inline one-line-cr-bot.sh:74]: customize the analysis *options* of the *tools used*,

File: one-line-cr-bot.sh [hunk 17]
```diff
@@ -0,0 +1,531 @@
...
+setup_cppcheck() {
+
+    local -i ret=0
+    local INSTALL_DIR="$1"
+    pushd "$INSTALL_DIR" 1>&2
+
+    echo "Setting up CppCheck..." 1>&2
+
+    # get cppcheck source
+    git --depth 1 -b 2.2 https://github.com/danmar/cppcheck.git || ret=$?
```
Comment (tautschnig) [inline one-line-cr-bot.sh:222]: In keeping with the approach for Infer, I'd go with downloading `https://github.com/danmar/cppcheck/archive/2.2.tar.gz` (and verify the checksum).

**Review by tautschnig — APPROVED** [review]

File: configuration/backends/infer/infer-evaluate.sh [hunk 18]
```diff
@@ -107,8 +107,10 @@ function evaluate_infer() {
   # run analysis on combined output from "infer capture --continue" calls
   echo "Running infer analyze (storing output in "$RESULTS_DIR"/infer_analyze.log) ..."
   local -i INFER_ANALYSIS_STATUS=0
+  INFER_ANALYSIS_EXTRA_ARGS="${INFER_ANALYSIS_EXTRA_ARGS:-}"
```
[the remark this answers is not in the document]
Comment (nmanthey) [reply configuration/backends/infer/infer-evaluate.sh:110]: It's mentioned below.

File: configuration/doc/EXAMPLES.md [hunk 19]
```diff
@@ -26,6 +26,17 @@ To store log files and results in another directory, the -o flag can be used:
 
  one-line-scan --no-analysis -o OLS -- make
 
+To add another run to the same output directory, depending on whether you want
+to keep the files from the previous run or not. To keep them, add the flag
```
[the remark this answers is not in the document]
Comment (nmanthey) [reply configuration/doc/EXAMPLES.md:30]: Will fix.

File: configuration/doc/EXAMPLES.md [hunk 20]
```diff
@@ -60,3 +60,23 @@ x86_64-unknown-linux-gcc can be used, as well as other cross-compilers. To use
 such a compiler, the --prefix parameter can be added.
 
  one-line-scan --fortify --no-gotocc --prefix x86_64-unknown-linux- -- make
+
+## Infer Backend
+
+For some project setups, running Infer can be challenging, for example for
+projects that use cmake. Although cmake allows to dump a compilation database,
```
[the remark this answers is not in the document]
Comment (nmanthey) [reply configuration/doc/EXAMPLES.md:67]: Agreed. Will drop and rewrite differently.

File: configuration/doc/EXAMPLES.md (hunk 1 above)
Comment (nmanthey) [reply configuration/doc/EXAMPLES.md:81]: export works, but would modify state.

File: configuration/doc/README.md (hunk 2 above)
Comment (nmanthey) [reply configuration/doc/README.md:40]: Yes, that would work for me.

File: one-line-cr-bot.sh (hunk 17 above)
Comment (nmanthey) [reply one-line-cr-bot.sh:222]: Fair point, will adapt.
